playwright-junit/src/main/java/com/microsoft/playwright/PlaywrightExtension.java

```diff
@@ -0,0 +1,50 @@
+package com.microsoft.playwright;
+
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+
+import static com.microsoft.playwright.ExtensionUtils.hasProperAnnotation;
+
+public class PlaywrightExtension implements ParameterResolver {
+  private static final ThreadLocal<Playwright> playwrightThreadLocal;
+
+  static {
+    playwrightThreadLocal = new ThreadLocal<>();
+  }
+
+  @Override
+  public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext) throws ParameterResolutionException {
+    if (!hasProperAnnotation(extensionContext)) {
+      return false;
+    }
+    Class<?> clazz = parameterContext.getParameter().getType();
+    return Playwright.class.equals(clazz);
+  }
+
+  @Override
+  public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext) throws ParameterResolutionException {
+    return getOrCreatePlaywright();
+  }
+
+  public static Playwright getOrCreatePlaywright() {
+    if (playwrightThreadLocal.get() == null) {
+      playwrightThreadLocal.set(Playwright.create());
+      System.out.println("Created Playwright " + getPlaywright());
+    }
+    return getPlaywright();
+  }
+
+  public static Playwright getPlaywright() {
+    return playwrightThreadLocal.get();
+  }
+
+  public static void closePlaywright() {
```

Comment:
Ideally we'd keep `Playwright` instances until all the tests finished running. Something like [RunListener.testRunFinished](https://junit.org/junit4/javadoc/4.12/org/junit/runner/notification/RunListener.html#testRunFinished(org.junit.runner.Result)) seems a good hook to run such logic but it looks like that interface was removed in JUnit 5. Maybe there is a new counterpart?

Reply:
Junit 5 doesn't have a built in after test run hook.

They have a workaround using a closeable resource but it currently could run in a separate thread than the class which means we couldn't access the threadlocal objects.

I've opened an issue with JUnit for this also.

For now, I think after class is our best option.